### DISCUSSION TITLE
Equatable CasePath

### DISCUSSION THREAD

**Discussion by mycroftcanner** [pointfreeco/swift-case-paths#26] [Ideas]
Is it possible to make CasePath Equatable?

**Comment by stephencelis** [comment]
Unfortunately not. While it would be possible to add an `identifier` field to equate against, it'd be a breaking change and there's no way as far as I can tell to automatically derive an identifier from an enum case embed function. Open to ideas! But it might not be possible and all the more reason to introduce case paths at the language level so they can be hashable like key paths.

Alternatively, you could wrap case paths in a type that _is_ equatable by providing the strategy for comparing paths. Not ideal, but could unblock you.

**Reply by tgrapperon** [reply]
> there's no way as far as I can tell to automatically derive an identifier from an enum case embed function

Is this still the case given latest developments from @mayoff  and yourself to the library?

`Equatable` (or better, `Hashable`) `CasePath`'s for enums could potentially open the route toward `Store` and `ViewStore` instances being able to be reused when scoping using `KeyPath`/`CasePath` or when "ViewStore-ing" in TCA.

**Reply by mayoff** [reply]
Sorry, but the latest developments don't make it easier to conform to `Equatable`.

**Reply by tgrapperon** [reply]
I get it now. All your work happens on the extraction side, but equatability should happen at rest. 
Thanks for the answer, and also for your incredible contributions.

**Comment by stephencelis** [comment]
I'm going to move this to be a discussion since it's not a bug, but is something that would be nice to solve (along with hashability), if anyone figures out how to do this!

**Comment by acosmicflamingo** [comment]
Oh no!!! Could CasePath's lack of Equatable & Hashable conformance be why my tests aren't detecting any changes in TestStore's state? In the code below, `updateSwitchRow` applies expected change in state, and it even works perfectly in the live app. This makes me think that the problem definitely comes from lack of conformance :(

```swift
let isOn = true
// item.action is of type CasePath<Reducer.Action, (UUID, Tagged<SwitchRowItem, UUID>, Bool)>
let action: UICollectionViewHelpersFeature.Action = item.action.embed((sectionId, itemId, isOn))
store.send(action) {
  $0.updateSwitchRow(sectionId: sectionId, itemId: itemId, isOn: isOn) // complains that state didn't change
}
```

**Reply by stephencelis** [reply]
All the more reason to keep pushing for support in the language ;) If case paths were in Swift proper, they'd have all the same affordances as key paths, including hashability and equitability.

For your purposes, though, you may be able to manually conform "item" to be equatable by using concrete `SwitchRowItem` instances to detect if the case is the same.

**Reply by acosmicflamingo** [reply]
@stephencelis thanks! I tried making some tweaks to what SwitchRowItem conforms to, and the tests pass! But, when I revert those changes, the test still passes!!! Turns out this has nothing to do with CasePaths not being equatable, and I have absolutely no idea why the test was failing yesterday :( I suppose I should be happy that it passes, but I am dumbfounded at what the problem was LoL

Edit: perhaps the problem was debugging late at night when I should be sleeping ;)
